### PR TITLE
ISPN-12178 Query broadcast blocks on the non-blocking executor

### PR DESCRIPTION
**Pull request by gustavonalle**
https://issues.redhat.com/browse/ISPN-12178

### CODE REVIEW

**Comment by gustavonalle** [comment]
I can't reproduce them locally, but it solved the ThreadLeaks issues on CI

**Comment by tristantarrant** [comment]
@wburns should look at this

File: query/src/main/java/org/infinispan/query/clustered/ClusteredQueryInvoker.java [hunk 1]
```diff
@@ -101,9 +72,8 @@ QueryResponse unicast(Address address, SegmentsClusteredQueryCommand cmd) {
 
       List<QueryResponse> results = new ArrayList<>();
       try {
-         results.add(localResponse.get());
-         List<QueryResponse> responseList = sequence(futureRemoteResponses).get();
-         results.addAll(responseList);
+         results.add(await(localResponse.toCompletableFuture()));
```
Comment (wburns) [inline query/src/main/java/org/infinispan/query/clustered/ClusteredQueryInvoker.java:75]: I worry that the blocking is now just relegated to this call instead. We really need https://issues.redhat.com/browse/ISPN-11731 to verify these are all fixed.

File: query/src/main/java/org/infinispan/query/clustered/commandworkers/CQCreateEagerQuery.java [hunk 2]
```diff
@@ -18,38 +18,35 @@
...
-         return null;
-      }
-
-      Object[] keys = queryResult.hits().stream()
-            .map(hit -> hit.id())
-            .map(id -> stringToKey(id))
-            .toArray(Object[]::new);
-      return new NodeTopDocs(cache.getRpcManager().getAddress(), queryResult.topDocs(), keys, null);
+   private CompletionStage<NodeTopDocs> collectKeys(SearchQueryBuilder query) {
+      return blockingManager.supplyBlocking(() -> query.documentReference().fetchAll(), this)
```
Comment (wburns) [inline query/src/main/java/org/infinispan/query/clustered/commandworkers/CQCreateEagerQuery.java:31]: I would recommend putting something a bit more descriptive for the second argument, maybe `CQCreateEagerQuery#collectKeys` ?

File: query/src/main/java/org/infinispan/query/clustered/commandworkers/CQCreateEagerQuery.java [hunk 3]
```diff
@@ -18,38 +18,35 @@
...
+            });
    }
 
-   private NodeTopDocs collectProjections(SearchQueryBuilder query) {
-      LuceneSearchResult<?> queryResult = query.build().fetchAll();
-      if (queryResult.totalHitCount() == 0L) {
-         return null;
-      }
+   private CompletionStage<NodeTopDocs> collectProjections(SearchQueryBuilder query) {
+      return blockingManager.supplyBlocking(() -> query.build().fetchAll(), this).thenApply(queryResult -> {
```
Comment (wburns) [inline query/src/main/java/org/infinispan/query/clustered/commandworkers/CQCreateEagerQuery.java:44]: Same here about second argument.

File: query/src/main/java/org/infinispan/query/clustered/ClusteredQueryInvoker.java (hunk 1 above)
Comment (gustavonalle) [reply query/src/main/java/org/infinispan/query/clustered/ClusteredQueryInvoker.java:75]: agreed, but this method is called by the REST server when doing queries, shouldn't Blockhound bark?

Comment (wburns) [reply query/src/main/java/org/infinispan/query/clustered/ClusteredQueryInvoker.java:75]: My guess is that the invoker is already on the blocking thread. But I am not certain what code path this is.

File: query/src/main/java/org/infinispan/query/clustered/ClusteredQueryInvoker.java [hunk 4]
```diff
@@ -116,26 +86,8 @@ QueryResponse unicast(Address address, SegmentsClusteredQueryCommand cmd) {
       return results;
    }
 
-   private Future<QueryResponse> localInvoke(SegmentsClusteredQueryCommand cmd) {
-      return asyncExecutor.submit(() -> {
```
Comment (wburns) [inline query/src/main/java/org/infinispan/query/clustered/ClusteredQueryInvoker.java:120]: Also noticed that it seems the `asyncExecutor` can be removed from this code and subsequent callers.

**Comment by wburns** [comment]
Integrated into master, thanks @gustavonalle !

**Pull request merged**